src/counters.rs

```diff
@@ -0,0 +1,178 @@
+    Strong,
+}
+
+type CounterVal = i64;
+
+pub struct CreateCounterReq<'i> {
+    infinispan_client: &'i Infinispan,
+    name: String,
+    counter_type: CounterType,
+    value: Option<CounterVal>,
+}
+
+impl<'i> CreateCounterReq<'i> {
+    pub fn new(
+        infinispan_client: &'i Infinispan,
+        name: impl Into<String>,
+        counter_type: CounterType,
+    ) -> CreateCounterReq<'i> {
+        CreateCounterReq {
+            infinispan_client,
+            name: name.into(),
+            counter_type,
+            value: None,
+        }
+    }
+
+    pub fn with_value(mut self, value: CounterVal) -> CreateCounterReq<'i> {
+        self.value = Some(value);
+        self
+    }
+
+    pub async fn send(&self) -> Result<Response, reqwest::Error> {
+        let counter = match self.counter_type {
+            CounterType::Weak => Counter::Weak(WeakCounter {
+                initial_value: self.value,
+            }),
+            CounterType::Strong => Counter::Strong(StrongCounter {
+                initial_value: self.value,
+            }),
+        };
+
+        let http_request = Request::post(self.infinispan_client.counter_url(&self.name))
+            .header(
+                AUTHORIZATION,
+                &self.infinispan_client.basic_auth_encoded_val,
+            )
+            .body(json!(counter).to_string())
+            .unwrap();
+
+        InfinispanRequest::new(http_request, self.infinispan_client.http_client.clone())
+            .send()
+            .await
+    }
+}
+
+pub struct IncrementCounterReq<'i> {
+    infinispan_client: &'i Infinispan,
+    name: String,
+    delta: Option<CounterVal>,
+}
+
+impl<'i> IncrementCounterReq<'i> {
+    pub fn new(
+        infinispan_client: &'i Infinispan,
+        name: impl Into<String>,
+    ) -> IncrementCounterReq<'i> {
+        IncrementCounterReq {
+            infinispan_client,
+            name: name.into(),
+            delta: None,
+        }
+    }
+
+    pub fn by(mut self, delta: CounterVal) -> IncrementCounterReq<'i> {
+        self.delta = Some(delta);
+        self
+    }
+
+    pub async fn send(&self) -> Result<Response, reqwest::Error> {
+        let http_request = Request::post(self.query_with_args())
+            .header(
+                AUTHORIZATION,
+                &self.infinispan_client.basic_auth_encoded_val,
+            )
+            .body("".into())
+            .unwrap();
+
+        InfinispanRequest::new(http_request, self.infinispan_client.http_client.clone())
+            .send()
+            .await
+    }
+
+    fn query_with_args(&self) -> String {
+        let mut query = format!(
+            "{}?action=add",
+            self.infinispan_client.counter_url(&self.name)
+        );
+
+        if let Some(delta) = self.delta {
+            query.push_str(&format!("&delta={}", delta));
+        }
+
+        query
+    }
+}
+
+impl Infinispan {
+    pub fn create_weak_counter(&self, name: impl Into<String>) -> CreateCounterReq {
```

Comment:
This kind of API will require &self to be spread everywhere while you create requests... doing so means the specific client must be alive no matter where in your code (or in different threads) you are building requests, plus that client can't have any &mut reference. Building requests should be independent of the client, imo.

Reply:
I've seen both styles in different rust libs. I pushed a branch with an different solution: https://github.com/3scale-labs/infinispan-rs/compare/alternative-first-version

The solution in that branch is more testable, there's less coupling, but it makes the API a bit uglier. I think the benefits of this branch outweigh the cons. Let me know what you think. I've only implemented the counters part for now, so ignore `Entries`and `Caches`.

It's used like this:
```rust
let infinispan = Infinispan::new("http://127.0.0.1:11222", "username", "password");
infinispan
  .run(InfinispanReq::create_weak_counter("some_counter").with_value(10))
  .await
  .unwrap();
```

Reply:
Can you rebase that branch on top of this one? And maybe open a draft PR on top of this one targeting this PR's branch so we can more easily see the changes and discuss there?

Reply:
I personally prefer the less coupled solution, and you might be able to emulate the coupled one on top of it.

Reply:
Pushed the changes in this PR :+1: